File: detective/evaluator.py
import os
import json
import urllib.request

TRANSCRIPTS_DIR = "transcripts"

API_KEY = os.environ.get("GEMINI_API_KEY")

if not API_KEY:
    raise ValueError("Please set GEMINI_API_KEY in the terminal.")


def load_prompt():
    base_dir = os.path.dirname(__file__)
    prompt_path = os.path.join(base_dir, "judge_prompt.md")

    with open(prompt_path, "r", encoding="utf-8") as f:
        return f.read()


SYSTEM_PROMPT = load_prompt()


def call_gemini(prompt):
    data = {
        "contents": [
            {
                "parts": [
                    {"text": prompt}
                ]
            }
        ]
    }

    req = urllib.request.Request(
        "https://generativelanguage.googleapis.com/v1beta/models/gemini-3-flash-preview:generateContent",
        data=json.dumps(data).encode("utf-8"),
        headers={
            "Content-Type": "application/json",
            "x-goog-api-key": API_KEY
        },
        method="POST"
    )

    with urllib.request.urlopen(req) as response:
        result = json.loads(response.read())

    text = result["candidates"][0]["content"]["parts"][0]["text"]
    return text


# ---------------------------------------------------------------------------
# Scoring
# ---------------------------------------------------------------------------
# Phase penalties
PHASE_PENALTIES = {
    "opening":     15,
    "discovery":   20,
    "negotiation": 25,
    "closing":     15,
}

# Tone violation penalty (per violation)
TONE_PENALTY = 10

# Critical failure penalty (per failure)
CRITICAL_PENALTY = 20


def compute_score(judge):
    score = 100

    # Phase failures
    for phase, penalty in PHASE_PENALTIES.items():
        if not judge.get(phase, {}).get("passed", True):
            score -= penalty

    # Tone violations
    score -= len(judge.get("tone_violations", [])) * TONE_PENALTY

    # Critical failures (new — each one costs 20 pts)
    score -= len(judge.get("critical_failures", [])) * CRITICAL_PENALTY

    score = max(0, min(score, 100))

    verdict = "good" if score >= 60 else "bad"

    return score, verdict


# ---------------------------------------------------------------------------
# Per-file evaluation
# ---------------------------------------------------------------------------

def judge_transcript(transcript):
    prompt = f"""{SYSTEM_PROMPT}

Evaluate this transcript:

{transcript}

Return JSON only."""

    response_text = call_gemini(prompt)

    # Strip markdown code fences if present
    cleaned = response_text.strip()
    if cleaned.startswith("```"):
        cleaned = cleaned.split("```", 2)[1]
        if cleaned.startswith("json"):
            cleaned = cleaned[4:]
        cleaned = cleaned.rsplit("```", 1)[0]
        cleaned = cleaned.strip()

    try:
        return json.loads(cleaned)
    except json.JSONDecodeError:
        print("Model returned invalid JSON:")
        print(response_text)
        raise


# ---------------------------------------------------------------------------
# Main loop
# ---------------------------------------------------------------------------

def evaluate_all():
    results = {}

    # Files to skip — not real transcripts
    SKIP_FILES = {"_manifest.json", "manifest.json"}

    for file in sorted(os.listdir(TRANSCRIPTS_DIR)):

        if file in SKIP_FILES:
            print(f"\nSkipping {file} (manifest — not a transcript)")
            continue

        path = os.path.join(TRANSCRIPTS_DIR, file)

        with open(path, "r", encoding="utf-8") as f:
            transcript = f.read()

        print(f"\nEvaluating {file}...")

        try:
            judge = judge_transcript(transcript)
            score, verdict = compute_score(judge)

            worst = [
                f'"{w["message"]}" — {w["reason"]}'
                for w in judge.get("worst_messages", [])
            ]

            results[file] = {
                "score": score,
                "worst_messages": worst,
                "verdict": verdict,
            }

            print(f"  Score:   {score}")
            print(f"  Verdict: {verdict}")

            crit = judge.get("critical_failures", [])
            if crit:
                print(f"  Critical failures ({len(crit)}):")
                for c in crit:
                    print(f"    [{c.get('type')}] {c.get('reason','')}")

        except Exception as e:
            print(f"  Failed to evaluate {file}: {e}")
            results[file] = {"error": str(e)}

    output_path = "evaluation_results.json"
    with open(output_path, "w", encoding="utf-8") as f:
        json.dump(results, f, indent=2, ensure_ascii=False)

    print(f"\nResults saved to {output_path}")
    return results


if __name__ == "__main__":
    evaluate_all()

def evaluate_call(conversation):

    transcript = []

    for turn in conversation:
        transcript.append({
            "speaker": "customer",
            "text": turn["customer"]
        })

        transcript.append({
            "speaker": "agent",
            "text": turn["agent"]
        })

    transcript_json = json.dumps({"transcript": transcript}, indent=2)

    judge = judge_transcript(transcript_json)

    score, verdict = compute_score(judge)

    issues = [
        f'{w["message"]} — {w["reason"]}'
        for w in judge.get("worst_messages", [])
    ]

    return score, issues, verdict

File: detective/evaluator.js
import fs from "fs";
import path from "path";
import process from "process";
import { fileURLToPath } from "url";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const TRANSCRIPTS_DIR = "transcripts";

const API_KEY = process.env.GEMINI_API_KEY;

if (!API_KEY) {
  throw new Error("Please set GEMINI_API_KEY in the terminal.");
}

// ---------------------------------------------------------------------------
// Load Prompt
// ---------------------------------------------------------------------------

function loadPrompt() {
  const promptPath = path.join(__dirname, "judge_prompt.md");
  return fs.readFileSync(promptPath, "utf-8");
}

const SYSTEM_PROMPT = loadPrompt();

// ---------------------------------------------------------------------------
// Gemini Call
// ---------------------------------------------------------------------------

async function callGemini(prompt) {
  const data = {
    contents: [
      {
        parts: [{ text: prompt }],
      },
    ],
  };

  const response = await fetch(
    "https://generativelanguage.googleapis.com/v1beta/models/gemini-3-flash-preview:generateContent",
    {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "x-goog-api-key": API_KEY,
      },
      body: JSON.stringify(data),
    }
  );

  const result = await response.json();

  const text =
    result.candidates?.[0]?.content?.parts?.[0]?.text ?? "";

  return text;
}

// ---------------------------------------------------------------------------
// Scoring
// ---------------------------------------------------------------------------

const PHASE_PENALTIES = {
  opening: 15,
  discovery: 20,
  negotiation: 25,
  closing: 15,
};

const TONE_PENALTY = 10;
const CRITICAL_PENALTY = 20;

function computeScore(judge) {
  let score = 100;

  for (const [phase, penalty] of Object.entries(PHASE_PENALTIES)) {
    if (!judge?.[phase]?.passed) {
      score -= penalty;
    }
  }

  score -= (judge?.tone_violations?.length || 0) * TONE_PENALTY;
  score -= (judge?.critical_failures?.length || 0) * CRITICAL_PENALTY;

  score = Math.max(0, Math.min(score, 100));

  const verdict = score >= 60 ? "good" : "bad";

  return [score, verdict];
}

// ---------------------------------------------------------------------------
// Transcript Judge
// ---------------------------------------------------------------------------

async function judgeTranscript(transcript) {
  const prompt = `${SYSTEM_PROMPT}

Evaluate this transcript:

${transcript}

Return JSON only.`;

  const responseText = await callGemini(prompt);

  let cleaned = responseText.trim();

  if (cleaned.startsWith("```")) {
    cleaned = cleaned.split("```", 3)[1];
    if (cleaned.startsWith("json")) {
      cleaned = cleaned.slice(4);
    }
    cleaned = cleaned.split("```")[0].trim();
  }

  try {
    return JSON.parse(cleaned);
  } catch (err) {
    console.log("Model returned invalid JSON:");
    console.log(responseText);
    throw err;
  }
}

// ---------------------------------------------------------------------------
// Evaluate All Files
// ---------------------------------------------------------------------------

export async function evaluateAll() {
  const results = {};

  const SKIP_FILES = new Set(["_manifest.json", "manifest.json"]);

  const files = fs.readdirSync(TRANSCRIPTS_DIR).sort();

  for (const file of files) {
    if (SKIP_FILES.has(file)) {
      console.log(`\nSkipping ${file} (manifest — not a transcript)`);
      continue;
    }

    const filePath = path.join(TRANSCRIPTS_DIR, file);
    const transcript = fs.readFileSync(filePath, "utf-8");

    console.log(`\nEvaluating ${file}...`);

    try {
      const judge = await judgeTranscript(transcript);
      const [score, verdict] = computeScore(judge);

      const worst = (judge.worst_messages || []).map(
        (w) => `"${w.message}" — ${w.reason}`
      );

      results[file] = {
        score,
        worst_messages: worst,
        verdict,
      };

      console.log(`  Score:   ${score}`);
      console.log(`  Verdict: ${verdict}`);

      const crit = judge.critical_failures || [];

      if (crit.length) {
        console.log(`  Critical failures (${crit.length}):`);
        for (const c of crit) {
          console.log(`    [${c.type}] ${c.reason || ""}`);
        }
      }
    } catch (e) {
      console.log(`  Failed to evaluate ${file}: ${e}`);
      results[file] = { error: String(e) };
    }
  }

  const outputPath = "evaluation_results.json";

  fs.writeFileSync(outputPath, JSON.stringify(results, null, 2));

  console.log(`\nResults saved to ${outputPath}`);

  return results;
}

// ---------------------------------------------------------------------------
// Evaluate Conversation
// ---------------------------------------------------------------------------

export async function evaluateCall(conversation) {
  const transcript = [];

  for (const turn of conversation) {
    transcript.push({
      speaker: "customer",
      text: turn.customer,
    });

    transcript.push({
      speaker: "agent",
      text: turn.agent,
    });
  }

  const transcriptJson = JSON.stringify({ transcript }, null, 2);

  const judge = await judgeTranscript(transcriptJson);

  const [score, verdict] = computeScore(judge);

  const issues = (judge.worst_messages || []).map(
    (w) => `${w.message} — ${w.reason}`
  );

  return { score, issues, verdict };
}

// ---------------------------------------------------------------------------
// CLI
// ---------------------------------------------------------------------------

if (process.argv[1] === __filename) {
  evaluateAll();
}